src/System.Management.Automation/engine/parser/Compiler.cs

```diff
@@ -955,33 +955,95 @@ private IEnumerable<Expression> CompileInvocationArguments(IEnumerable<Expressio
             return arguments == null ? Array.Empty<Expression>() : arguments.Select(CompileExpressionOperand);
         }
 
-        internal Expression ReduceAssignment(ISupportsAssignment left, TokenKind tokenKind, Expression right)
+        internal Expression ReduceAssignment(
+            ISupportsAssignment left,
+            TokenKind tokenKind,
+            Expression right,
+            MemberExpressionAst lhsMemberAccessAst)
         {
-            IAssignableValue av = left.GetAssignableValue();
-            ExpressionType et = ExpressionType.Extension;
+            IAssignableValue assignmentTarget = left.GetAssignableValue();
 
-            switch (tokenKind)
+            if (tokenKind == TokenKind.Equals)
             {
-                case TokenKind.Equals: return av.SetValue(this, right);
-                case TokenKind.PlusEquals: et = ExpressionType.Add; break;
-                case TokenKind.MinusEquals: et = ExpressionType.Subtract; break;
-                case TokenKind.MultiplyEquals: et = ExpressionType.Multiply; break;
-                case TokenKind.DivideEquals: et = ExpressionType.Divide; break;
-                case TokenKind.RemainderEquals: et = ExpressionType.Modulo; break;
-                case TokenKind.QuestionQuestionEquals: et = ExpressionType.Coalesce; break;
+                return assignmentTarget.SetValue(this, right);
             }
 
+            var expressionType = tokenKind switch
+            {
+                TokenKind.PlusEquals => ExpressionType.Add,
+                TokenKind.MinusEquals => ExpressionType.Subtract,
+                TokenKind.MultiplyEquals => ExpressionType.Multiply,
+                TokenKind.DivideEquals => ExpressionType.Divide,
+                TokenKind.RemainderEquals => ExpressionType.Modulo,
+                TokenKind.QuestionQuestionEquals => ExpressionType.Coalesce,
+                _ => ExpressionType.Extension
+            };
+
+            return ReduceCompoundAssignment(this, assignmentTarget, lhsMemberAccessAst, expressionType, right);
+        }
+
+        internal static BlockExpression ReduceCompoundAssignment(
+            Compiler compiler,
+            IAssignableValue assignmentTarget,
+            MemberExpressionAst lhsMemberExpressionAst,
+            ExpressionType expressionType,
+            Expression valueToAssign)
+        {
             var exprs = new List<Expression>();
             var temps = new List<ParameterExpression>();
-            var getExpr = av.GetValue(this, exprs, temps);
+            var getExpr = assignmentTarget.GetValue(compiler, exprs, temps);
 
-            if (et == ExpressionType.Coalesce)
+            if (expressionType == ExpressionType.Coalesce)
             {
-                exprs.Add(av.SetValue(this, Coalesce(getExpr, right)));
+                exprs.Add(assignmentTarget.SetValue(compiler, Coalesce(getExpr, valueToAssign)));
+            }
+            else if (lhsMemberExpressionAst != null
```

Comment:
This suggests that this will only work when events are given in the form:

```powershell
$x.Event += $handler
```

But in the comment below you give an alternate form:

```powershell
$e = $x.Event
$e += $handler
```

Should we expect that to behave differently? If so, why? Is that consistent with other similar behaviours in PowerShell today?

Reply:
Hmm, that's a good point! I'm not sure, I guess I'd expect that that behaves the same. 🤔 

I was mostly mimicking C#'s syntax, not sure saving the event to a var works there, but I guess I'd expect that works also.

Reply:
Looking back at this, the example you're referring to is more illustrating how the `$object.EventName += $sb` expression is compiled (it's saved to a temporary variable first) rather than what would literally work.

Do we want to allow this to work for PSEvent instances saved in a variable as well? My concern here was mainly that the `$object.EventName += $b` is likely to be the most common usage, and I didn't want to bog down _every_ `$var += $sb` expression with a runtime check for `$var -is [PSEvent]` if I could avoid it.

Reply:
Looking at Sharplab, I can see the C# compiler actually completely prevents you assigning an event handler reference to a variable in the first place:
![image](https://user-images.githubusercontent.com/32407840/90328278-1bd93000-df69-11ea-91b2-a830f51dc00c.png)

I don't think we need to be _quite_ so strict in PS, but I think folks trying to use `+=` / `-=` for event handling will expect similar behaviour.

Reply:
Right, absolutely. I actually think you're right, but wanted to ask the question. (I referred to the compilation example because I knew it was a compilation example, but it *was* why I thought of it).

Events are strange second-class citizens in C# (I don't believe "event handler reference" is a meaningful concept for C#). But with that said, I'm not sure how many people we help by making them first-class in PowerShell (at the cost of things like runtime performance).

Reply:
👍 for how it's currently implemented. It's consistent with C# and also with how `PSMethod`'s are handled.   Like you can `$method = $obj.ToString` but you can't then do `$method()`